### ISSUE TITLE
Casks: "Backing App (app) up to (path)" step can fail due to missing read permissions inside an application bundle

### ISSUE THREAD

**Issue by charlieok** [Homebrew/brew#16536]
### `brew doctor` output

```shell
homebrew_admin@Cs-M2-MacBook-Pro ~ % brew doctor
Your system is ready to brew.
```

### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```shell
homebrew_admin@Cs-M2-MacBook-Pro ~ % brew config
HOMEBREW_VERSION: 4.2.5
ORIGIN: https://github.com/Homebrew/brew
HEAD: 9c8374d3b78c4f531c5c2c0d272d1fb1f1b60d70
Last commit: 5 days ago
Core tap JSON: 26 Jan 15:54 UTC
Core cask tap JSON: 26 Jan 15:54 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: dodeca-core 64-bit arm_blizzard_avalanche
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.3-arm64
CLT: 15.1.0.0.1.1700200546
Xcode: N/A
Rosetta 2: false
```

### What were you trying to do (and why)?

force-uninstall the 'signal' cask

why: I actually want to upgrade the cask, and ran into an error, possibly related to this issue. The next thing I tried to do was uninstall the cask, which is what this issue is about.

### What happened (include all command output)?

```
homebrew_admin@Cs-M2-MacBook-Pro ~ % brew uninstall --cask signal --force  --debug
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading signal
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromPathLoader): loading /opt/homebrew/Caskroom/signal/.metadata/6.41.0/20231207053038.799/Casks/signal.json
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading tree from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading libtiff from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading binaryen from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading little-cms2 from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading libpng from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading ncurses from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading ghostscript from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading freetype from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading rustup-init from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading openjpeg from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading lz4 from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading elvish from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading xz from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading fontconfig from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading openssl@3 from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading ca-certificates from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading zstd from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading jbig2dec from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading pcre2 from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading direnv from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading libidn from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading fish from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading wabt from API
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading jpeg-turbo from API
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading coderunner
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading db-browser-for-sqlite
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading firefox
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading gitup
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading google-chrome
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading little-snitch
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading obsidian
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading protonvpn
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading rancher
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading signal
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading sourcetree
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading sublime-merge
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading sublime-text
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading tor-browser
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading visual-studio-code
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading vlc
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading wireshark
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading zoom-for-it-admins
==> Uninstalling Cask signal
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromPathLoader): loading /opt/homebrew/Caskroom/signal/.metadata/6.41.0/20231207053038.799/Casks/signal.json
==> Uninstalling Cask signal
==> Uninstalling artifacts
==> 2 artifacts defined
#<Cask::ArtifactSet: {#<Cask::Artifact::App:0x000000010c95f440 @cask=#<Cask signal>, @dsl_args=["Signal.app"], @source_string="Signal.app", @target_string="">, #<Cask::Artifact::Zap:0x000000010c95e7e8 @cask=#<Cask signal>, @dsl_args=[{:trash=>["~/Library/Application Support/Signal", "~/Library/Preferences/org.whispersystems.signal-desktop.helper.plist", "~/Library/Preferences/org.whispersystems.signal-desktop.plist", "~/Library/Saved Application State/org.whispersystems.signal-desktop.savedState"]}], @directives={:trash=>["~/Library/Application Support/Signal", "~/Library/Preferences/org.whispersystems.signal-desktop.helper.plist", "~/Library/Preferences/org.whispersystems.signal-desktop.plist", "~/Library/Saved Application State/org.whispersystems.signal-desktop.savedState"], :signal=>[]}>}>
==> Uninstalling artifact of class Cask::Artifact::App
Warning: It seems there is already an App at '/opt/homebrew/Caskroom/signal/6.41.0/Signal.app'; overwriting.
==> Removing App '/opt/homebrew/Caskroom/signal/6.41.0/Signal.app'
==> Backing App 'Signal.app' up to '/opt/homebrew/Caskroom/signal/6.41.0/Signal.app'
/usr/bin/env /bin/cp -pR /Applications/Signal.app /opt/homebrew/Caskroom/signal/6.41.0/Signal.app
cp: readlink: /Applications/Signal.app/Contents/Frameworks/ReactiveObjC.framework/Resources: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/ReactiveObjC.framework/ReactiveObjC: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/ReactiveObjC.framework/Versions/Current: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/Squirrel.framework/Resources: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/Squirrel.framework/Versions/Current: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/Squirrel.framework/Squirrel: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/Mantle.framework/Mantle: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/Mantle.framework/Resources: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/Mantle.framework/Versions/Current: Permission denied
Error: Failure while executing; `/usr/bin/env /bin/cp -pR /Applications/Signal.app /opt/homebrew/Caskroom/signal/6.41.0/Signal.app` exited with 1. Here's the output:
cp: readlink: /Applications/Signal.app/Contents/Frameworks/ReactiveObjC.framework/Resources: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/ReactiveObjC.framework/ReactiveObjC: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/ReactiveObjC.framework/Versions/Current: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/Squirrel.framework/Resources: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/Squirrel.framework/Versions/Current: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/Squirrel.framework/Squirrel: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/Mantle.framework/Mantle: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/Mantle.framework/Resources: Permission denied
cp: readlink: /Applications/Signal.app/Contents/Frameworks/Mantle.framework/Versions/Current: Permission denied

/opt/homebrew/Library/Homebrew/system_command.rb:345:in `assert_success!'
/opt/homebrew/Library/Homebrew/system_command.rb:68:in `run!'
/opt/homebrew/Library/Homebrew/system_command.rb:33:in `run'
/opt/homebrew/Library/Homebrew/system_command.rb:37:in `run!'
/opt/homebrew/Library/Homebrew/cask/artifact/moved.rb:145:in `move_back'
/opt/homebrew/Library/Homebrew/cask/artifact/moved.rb:23:in `uninstall_phase'
/opt/homebrew/Library/Homebrew/cask/installer.rb:469:in `block in uninstall_artifacts'
/opt/homebrew/Library/Homebrew/cask/artifact_set.rb:14:in `each'
/opt/homebrew/Library/Homebrew/cask/artifact_set.rb:14:in `each'
/opt/homebrew/Library/Homebrew/cask/installer.rb:466:in `uninstall_artifacts'
/opt/homebrew/Library/Homebrew/cask/installer.rb:404:in `uninstall'
/opt/homebrew/Library/Homebrew/cask/uninstall.rb:15:in `block in uninstall_casks'
/opt/homebrew/Library/Homebrew/cask/uninstall.rb:10:in `each'
/opt/homebrew/Library/Homebrew/cask/uninstall.rb:10:in `uninstall_casks'
/opt/homebrew/Library/Homebrew/cmd/uninstall.rb:77:in `uninstall'
/opt/homebrew/Library/Homebrew/brew.rb:86:in `<main>'
```

I can see in the debug output that the specific command that caused the error was this one:

```
/usr/bin/env /bin/cp -pR /Applications/Signal.app /opt/homebrew/Caskroom/signal/6.41.0/Signal.app
```

The "Permission denied" lines show the reason: some directories underneath `/Applications/Signal.app/Contents/Frameworks/` have their non-owner permission bits unset. For example:

```
homebrew_admin@Cs-M2-MacBook-Pro ReactiveObjC.framework % ls -l /Applications/Signal.app/Contents/Frameworks/ReactiveObjC.framework
total 0

ls: /Applications/Signal.app/Contents/Frameworks/ReactiveObjC.framework/ReactiveObjC: Permission denied
lrwx------  1 root  wheel   29 Dec 14 15:59 ReactiveObjC

ls: /Applications/Signal.app/Contents/Frameworks/ReactiveObjC.framework/Resources: Permission denied
lrwx------  1 root  wheel   26 Dec 14 15:59 Resources
drwxr-xr-x@ 4 root  wheel  128 Dec 14 15:59 Versions
```

Looking at the source, I think it happens right here:
https://github.com/Homebrew/brew/blob/4.2.5/Library/Homebrew/cask/artifact/moved.rb#L145

That line is passing a 'sudo:' argument, which it will set if it finds that the target path for the cp command isn't writable. It looks to me like this could be fixed by adding a test checking whether the source path is readable. (But it would have to check recursively, since the non-readable bits were nested a few levels underneath the source path.)

### What did you expect to happen?

successful uninstall of the signal cask

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew install --cask signal
brew uninstall --cask signal --force  --debug
```

**Comment by charlieok** [comment]
I worked around this by manually changing the permissions of the contents of the application bundle to allow the command to complete. After installing the upgraded signal cask, simply installing and uninstalling it did not reproduce this issue for me.

But, knowing that I did wind up in a state with some files having root-only read permissions tells me that casked app bundles can sometimes get into that state. Manually removing the non-owner permission bits between installing and uninstalling should be sufficient to reproduce this issue.
